### ISSUE TITLE
lein pulls in Clojure 1.6 rather than 1.8

### ISSUE THREAD

**Issue by Reefersleep** [technomancy/leiningen#2084]
I discovered this while getting an error trying to do `lein new figwheel`. bhauman figured out that what caused the error was that somehow my lein was using Clojure 1.6 rather than 1.8, and lein-figwheel was using language features not in 1.6. bhauman changed lein-figwheel so that it supported 1.6, but gave a warning if it detected this version.

The issue where I reported this at the lein-figwheel repo is here: https://github.com/bhauman/lein-figwheel/issues/285

In short, I'm using lein 2.5.3 on a PC with Windows 8, and upgrading leiningen didn't help. I originally downloaded leiningen from leiningen.org . My ~\.lein\profiles.clj is empty.

**Comment by hyPiRion** [comment]
Could you explain the exact steps to reproduce this issue?

According to that thread, you do

```shell
lein new figwheel foobar
cd foobar
lein clean
lein figwheel
```

and it crashes with the error presented there, is that correct?

**Comment by Reefersleep** [comment]
Hello @hyPiRion 

Sorry for acting so unprofessional :) Just wanted to get the issue in before I forgot to report it at all. 

Yes, I do what you describe, and prior to bhauman making lein-figwheel Clojure 1.6 compatible and yogthos bumping lein-figwheel to that version, the `lein figwheel` would crash. Now, however, the result is as in the paste below.

```
c:\dev\clojure\Projects>lein new figwheel foobar
Retrieving figwheel/lein-template/0.5.0-7/lein-template-0.5.0-7.pom from clojars
Retrieving figwheel/lein-template/0.5.0-7/lein-template-0.5.0-7.jar from clojars
Generating fresh 'lein new' figwheel project.
c:\dev\clojure\Projects>cd foobar

c:\dev\clojure\Projects\foobar>lein clean
c:\dev\clojure\Projects\foobar>lein figwheel
System Warning: Detected Clojure Version 1.6.0
  Figwheel requires Clojure 1.7.0 at least.
  This may only be occuring in the Leinigen (bootstrapping) process but still something to be aware of.
  Especially if this message is immediately followed by an strange stack trace.
  Check lein deps :tree or lein deps :plugin-tree for clues.
  Also, don't forget the influence of profiles.clj
Figwheel: Starting server at http://localhost:3449
Figwheel: Watching build - dev
←[0mCompiling "resources/public/js/compiled/foobar.js" from ["src"]...
←[32mSuccessfully compiled "resources/public/js/compiled/foobar.js" in 8.9 seconds.←[0m
Figwheel: Starting CSS Watcher for paths  ["resources/public/css"]
Launching ClojureScript REPL for build: dev
Figwheel Controls:
          (stop-autobuild)                ;; stops Figwheel autobuilder
          (start-autobuild [id ...])      ;; starts autobuilder focused on optional ids
          (switch-to-build id ...)        ;; switches autobuilder to different build
          (reset-autobuild)               ;; stops, cleans, and starts autobuilder
          (reload-config)                 ;; reloads build config and resets autobuild
          (build-once [id ...])           ;; builds source one time
          (clean-builds [id ..])          ;; deletes compiled cljs target files
          (print-config [id ...])         ;; prints out build configurations
          (fig-status)                    ;; displays current state of system
  Switch REPL build focus:
          :cljs/quit                      ;; allows you to switch REPL to another build
    Docs: (doc function-name-here)
    Exit: Control+C or :cljs/quit
 Results: Stored in vars *1, *2, *3, *e holds last exception object
Prompt will show when Figwheel connects to your application
```

**Comment by hyPiRion** [comment]
@Reefersleep, oh, no worries – you don't act unprofessional at all. I just need to understand what's going on =)

What happens when you do `lein version`, and what does the `lein repl` prelude tell you? The repl prelude should be something like

```
nREPL server started on port 34905 on host 127.0.0.1 - nrepl://127.0.0.1:34905
REPL-y 0.3.7, nREPL 0.2.12
Clojure 1.8.0
OpenJDK 64-Bit Server VM 1.6.0_31-b31
```

**Comment by Reefersleep** [comment]
```
c:\dev\clojure\Projects\foobar>lein version
Leiningen 2.5.3 on Java 1.8.0_65 Java HotSpot(TM) 64-Bit Server VM
```
and
```
c:\dev\clojure\Projects\foobar>lein repl
nREPL server started on port 55201 on host 127.0.0.1 - nrepl://127.0.0.1:55201
REPL-y 0.3.5, nREPL 0.2.6
Clojure 1.7.0
Java HotSpot(TM) 64-Bit Server VM 1.8.0_65-b17
```

**Comment by hyPiRion** [comment]
It seems like there are still some user profiles on the path, according to the `lein repl` output: For 2.5.3, REPL-y should be on 0.3.7 and nREPL should be on 0.2.10.

You say `~\.lein\profiles.clj` (for reference, `%userprofile%` is `~` on Windows iirc) is empty ; is `~\.lein\profiles.d` a directory, and does it contain any files? If so, then the files in there set up profiles as well.

If not, check if the environment variable `LEIN_HOME` is set. If it is, then check the profiles there.

Finally, you could try to upgrade to 2.6.1 and see if that changes anything. I do not believe it does, unfortunately.

**Comment by Reefersleep** [comment]
To be accurate, `~\.lein\profiles.clj` contains the following:
```
{:user {:plugins []
        :dependencies []}}
```
There is no `~\.lein\profiles.d` dir.
I don't have a LEIN_HOME, only a LEIN_JAR.
I just upgraded to 2.6.1, but `lein figwheel` gives the same result.

**Comment by hyPiRion** [comment]
Well, then I sadly have absolutely no idea how this could be happening. =|

**Comment by Reefersleep** [comment]
Huh :\

Well, it's not a problem for me right now, and I'll hopefully be moving on to something unix-based soon, which I suspect will make the problem disappear entirely from my world. I hope that either I'm especially unfortunate in having experienced this, or that many other experience it and chime in so we can get to the bottom of it.

Thanks for your help, and thank you for Leiningen! :) :+1:

**Comment by hyPiRion** [comment]
Yeah, I hope there's some very weird configuration that causes this.

I'm closing this if you don't mind, as it doesn't seem to cause any big problems for now. However, feel free to reopen the moment this becomes an issue again (or post here if anyone else encounters this problem).

**Comment by Reefersleep** [comment]
Just want to note that if I do `lein new foobar` and read the project.clj, it lists as dependencies: `:dependencies [[org.clojure/clojure "1.6.0"]]`.

**Comment by hyPiRion** [comment]
Then you use an older version of Leiningen somehow – 2.5.3 shipped with 1.7.0 and the 2.6.x versions ships with 1.8.0 both in Lein proper and in the template.

**Comment by Reefersleep** [comment]
Apply :raised_hand: to :expressionless:

My LEIN_JAR environment variable was pointing towards a 2.5.0 version I had downloaded at some point. Removing the LEIN_JAR made everything ok - no warnings on `lein figwheel`, and `lein new foobar` uses Clojure 1.8 in the project.clj

I find it a bit weird though that when using `lein -v` and when upgrading and such, it never reported that it was using 2.5.0.

**Comment by hyPiRion** [comment]
yeah, that sounds very much like a bug. I'll have a look at that later today.

**Comment by Reefersleep** [comment]
Yeah, good job! :+1: 

And again, thank you for Leiningen! :D

**Comment by hyPiRion** [comment]
No problem, thanks for taking the time to investigate this with me =)

I'm still unable to figure out how it could say `lein 2.6.1` with LEIN_JAR set. I am on linux though, so I have a hard time verifying windows issues.

I think we should probably update the documentation around LEIN_JAR and recommend against using it, as it gives you a bad time when you're doing update.

**Comment by Reefersleep** [comment]
:) No problem.

I could give it a go with adding stuff to the documentation and perform a pull request, trying to keep it in the style of the rest of the documentation. Something like:

Windows

There is an installer which will handle downloading and placing Leiningen and its dependencies.

The manual method of putting the batch file. on your PATH and running lein self-install should still work for most users. If you have Cygwin you should be able to use the shell script above rather than the batch file.

N.B.: Avoid using a LEIN_JAR environment variable. If you have one instance of Leiningen on your PATH as well as a LEIN_JAR pointing to a different instance, there is a known problem where `lein -v` and `lein update` will refer to the first instance, while usage of Leiningen, such as `lein new foobar`, will refer to the latter.

**Comment by Reefersleep** [comment]
Just a suggestion. :)

By the way, I could spend a little time verifying the behaviour, just to be sure. We could also experiment further if you have any ideas.

**Comment by hyPiRion** [comment]
That seems like a fine inclusion to me – if you could also add in a small warning in [lein.bat](https://github.com/technomancy/leiningen/blob/0ce5e1ff87976637d8170e3c7f76a3c9e872820d/bin/lein.bat#L21) as well (saying it's deprecated or that you should only use this is you really know what you're doing), then that would be wonderful!

**Comment by Reefersleep** [comment]
I'll do that! :)

**Comment by Reefersleep** [comment]
I suppose that the lein.bat warning should go in a comment directly after the comment on line 21 saying
`:: LEIN_JAR and LEIN_HOME variables can be set manually.`
?
I mean, as opposed to `echo`ing it upon LEIN_JAR usage somewhere else.

**Comment by Reefersleep** [comment]
I'm thinking
```
:: LEIN_JAR and LEIN_HOME variables can be set manually.
:: Only set LEIN_JAR manually if you know what you are doing. Having LEIN_JAR pointing to one version of Leiningen as well as having a different version elsewhere in PATH has been known to cause problems.
```

**Comment by hyPiRion** [comment]
Yes, that sounds good =)

**Comment by hyPiRion** [comment]
Fixed via #2099.
